userdocs/diagnostics/expression-too-complex.md

```diff
@@ -0,0 +1,16 @@
+# Type Inference in Complex Expressions
+For some complex expressions, the Swift compiler is unable to infer types in reasonable time. This can occur for a number of reasons, but some common ones include:
+
+- The expression includes many overloaded operators and/or functions.
+- The expression contains literals which can be used as values of many different types (e.g. integer and floating point literals).
+- The expression includes a collection literal with a large number of nontrivial element expressions.
+
+These types of expressions can take a long time to typecheck because the compiler must consider and eliminate a larger number of possible solutions when assigning types to subexpressions.
```

Comment:
Some examples would probably be nice, though they'd become out of date if we can fix the exponential type checking issues.

Reply:
I'll see if I can come up with a reasonable example or two. I'd definitely like to avoid anything that might become out of date to avoid confusion, since every release tends to bring improvements in this area.

Reply:
We have a number of examples of flow type-checking in `validation-test/Sema/type_checker_perf/slow`.

Reply:
I added an example based on SR-12267 which is fairly realistic